i/b/log_observe: add CAP_SYSLOG for /dev/kmsg access on Jammy and above

alexclewontin opened the pull request:
As of Jammy, the default setting for `sysctl kernel.dmesg_restrict` changed from 0 -> 1. This means that `CAP_SYSLOG` is required for to access `/dev/kmsg`. Observably, on a 20.04 host (regardless of snap base), the log-observe plug was sufficient for a snap to run `dmesg` in confinement, but on a 22.04 host, it fails with message `dmesg: read kernel buffer failed: Operation not permitted` even with UID == 0 (snappy-debug output: `= AppArmor =
Time: Jun 26 16:29:49
Log: apparmor="DENIED" operation="capable" profile="snap.log-observe-test.test" pid=4030401 comm="dmesg" capability=34  capname="syslog"
Capability: syslog
`) Adding this snippet to the AA profile allows access.

This will need review from Security, as CAP_SYSLOG obviously has other implications.

Related: [SNAPDENG-35243](https://warthogs.atlassian.net/browse/SNAPDENG-35243)
                                                                 

[SNAPDENG-35243]: https://warthogs.atlassian.net/browse/SNAPDENG-35243?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

alexclewontin committed:
i/b/log_observe: add CAP_SYSLOG for /dev/kmsg access on Jammy and above

alexmurray reviewed (commented):
This is an interesting change - as per `capabilities(7)`:

>        CAP_SYSLOG (since Linux 2.6.37)
>              •  Perform  privileged  syslog(2) operations.  See syslog(2) for
>                 information on which operations require privilege.
>              •  View kernel addresses exposed via /proc and other  interfaces
>                 when  /proc/sys/kernel/kptr_restrict  has  the value 1.  (See
>                 the discussion of the kptr_restrict in proc(5).)

Looking then at `syslog(2)` we see:

>        All  commands  except 3 and 10 require privilege.  In Linux kernels be‐
>       fore Linux 2.6.37, command types 3 and 10 are allowed  to  unprivileged
>       processes;  since  Linux 2.6.37, these commands are allowed to unprivi‐
>       leged processes only if /proc/sys/kernel/dmesg_restrict has  the  value
>       0.   Before  Linux  2.6.37,  "privileged" means that the caller has the
>       CAP_SYS_ADMIN capability.  Since Linux 2.6.37, "privileged" means  that
>       the  caller has either the CAP_SYS_ADMIN capability (now deprecated for
>       this purpose) or the (new) CAP_SYSLOG capability.

Which is to say that since we changed this default the only way to get access to `/dev/kmsg` is to have `CAP_SYSLOG` - hence this PR.

However as can also be seen, this then allows the caller to perform many other operations like setting the console log level and clearing the ring buffer. The default seccomp template in snapd allows the use of the `syslog` system-call and so this would then allow any snap which plugs `log-observe` to now be able to clear the syslog buffer as well - as such this changes the semantics of this interface and I do not think it is appropriate to grant it.

Instead would it be better to use `journalctl` rather than directly reading `/dev/kmsg`?

alexclewontin commented:
@alexmurray appreciate the review, and everything you said makes sense.

To provide some context about the use case where this came up, it was caught in a "regression" of a third party log gathering app which had been packaged as a snap. I can look at whether that app can be configured to just gather this stuff via journald, but it may or may not be as simple as changing the approach of something completely homegrown.

Is there theoretically room for a more privileged `log-control` interface which would cover these types of operations?

alexmurray commented:
Yes I would prefer perhaps a new interface rather than changing the semantics of this interface - `log-control` sounds like a good approach (although @pedronis is the ultimate authority on this sort of thing)

bboozzoo commented:
@alexmurray could we maybe add some seccomp rules to block the 'write' like commands? eg:

```
~syslog SYSLOG_ACTION_CONSOLE_LEVEL - -
~syslog SYSLOG_ACTION_CONSOLE_OFF - -
~syslog SYSLOG_ACTION_CONSOLE_LEVEL - -
~syslog SYSLOG_ACTION_CONSOLE_ON - -
```

pedronis commented:
"-observe" suffixed interfaces are not meant to give write-like permissions, so either we introduce a new interface, or need to use seccomp filtering carefully to block the write-like functionality enabled by CAP_SYSLOG, doing that though we need to be careful of the interactions with other interfaces that might have given that out for the full functionality

alexclewontin commented:
I am happy to refactor this PR to instead introduce a new "log-control" interface, but want to hear the answer to @bboozzoo's question first, as it may mean it is not necessary for this use case.

alexmurray commented:
So I think the idea of using seccomp to limit the access to `syslog()` seems like a good, pragmatic solution - but it does make me a bit nervous that we may end up missing something and inadvertently turning this -observe interface into a -control interface. But then that means we just need to be thorough in the research for the implementation. So +1 from me for taking that route for now - FWIW we already take a similar approach in the base template where we allow the ioctl syscall but deny use of dangerous parameters.

If you choose to take this approach @alexclewontin can you please add a spread test that tests that a snap which plugs log-observe is denied from using `syslog()` with these different dangerous arguments? Thanks.

alexclewontin commented:
@alexmurray will do, appreciate the guidance! @pedronis I will take a crack at refactoring + adding spread test as per Alex's advice

alexclewontin committed:
cmd/snap-seccomp: filter syslog() calls by SYSLOG_ACTION_*

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

alexclewontin committed:
i/seccomp: explicitly deny "control"-type syslog() calls

This lets us let AppArmor allow CAP_SYSLOG in the log-observe interface
without giving up these more privileged accesses on Jammy+.

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

alexclewontin committed:
tests: test syslog(2) calls via spread

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

alexmurray commented on cmd/snap-seccomp/main_test.go:
```diff
@@ -847,6 +847,42 @@ func (s *snapSeccompSuite) TestRestrictionsWorkingArgsTermios(c *C) {
+	}{
+		// good input
+		{"syslog SYSLOG_ACTION_SIZE_UNREAD", "syslog;native;SYSLOG_ACTION_SIZE_UNREAD", Allow},
+		{"syslog SYSLOG_ACTION_CLOSE", "syslog;native;0", Allow},
```
Should this and the following `Allow` test cases use the named identifier rather than the 0,1,2 .. etc magic numbers?

alexclewontin replied:
Ah yes this is left over from one of my previous iterations, will tweak

alexmurray commented on tests/main/interfaces-log-observe/task.yaml:
```diff
@@ -8,11 +8,22 @@ details: |
+        echo "Ensure that on Jammy+, sysctl kernel.dmesg_restrict == 1"
+        # This should be the default, but we write it out just to be doubly sure
+        # On systems before Jammy, leave it as whatever the default is.
+        sysctl -w kernel.dmesg_restrict=1
```
Do we need to restore this back to the original value on `restore`?

alexclewontin replied:
Ah great catch. In practice it *probably* doesn't matter because I don't expect this to ever actually change the setting, but we definitely should just in case. I'll address

alexmurray reviewed (changes requested):
Thanks @alexclewontin - this is looking great - just a couple of minor comments.

alexclewontin committed:
many: clean up tests, typos

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

alexmurray reviewed (approved):
LGTM - thanks for all your work on this @alexclewontin.

One minor worry I have is that the denylist approach here means that any new syslog actions will be automatically allowed  - I wonder how gross it would be to preemptively deny all actions greater than SYSLOG_ACTION_SIZE_BUFFER - it might be doable with:

```
~syslog >SYSLOG_ACTION_SIZE_BUFFER
```

But then we would want to add a test for that as well then - thoughts?

pedronis commented:
@alexmurray 

> ~syslog >SYSLOG_ACTION_SIZE_BUFFER
>  But then we would want to add a test for that as well then - thoughts?

it seems to be on the safe side we do need to do that?

alexmurray commented:
@alexclewontin do you think you might be able to try adding the additional denylist entry for any future/unknown syslog actions as well as an additional test for this?

alexclewontin commented:
Yep, sorry for delay, I am adding

alexclewontin committed:
i/seccomp,tests: deny potential future syslog(2) calls

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

alexmurray reviewed (approved):
Thanks @alexclewontin - LGTM!

bboozzoo commented on tests/main/interfaces-log-observe/task.yaml:
```diff
@@ -44,3 +89,30 @@ execute: |
     fi
+
+    echo "Or make ANY privileged syslog(2) calls"
+    log-observe-consumer.klogctl SYSLOG_ACTION_CLOSE | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_OPEN | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_READ | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_READ_CLEAR | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_CLEAR | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_CONSOLE_OFF | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_CONSOLE_ON | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_CONSOLE_LEVEL | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_SIZE_UNREAD | MATCH 'ERROR (operation not permitted|permission denied)'
+    # It is important that the following matches EPERM/EACCES, not EINVAL, as it
+    # is a stand-in for potential future SYSLOG_ACTIONs
+    log-observe-consumer.klogctl SYSLOG_ACTION_BAD | MATCH 'ERROR (operation not permitted|permission denied)'
```
EACCESS here

alexclewontin replied:
SYSLOG_ACTION_READ and SYSLOG_ACTION_SIZE_UNREAD are denied with EPERM because they are allowed by seccomp but denied by lack of CAP_SYSLOG (similar to the SYSLOG_ACTION_READ_ALL and SYSLOG_ACTION_SIZE_BUFFER on dmesg_restrict systems), but will make sure each action is targeting the appropriate error

bboozzoo commented on tests/main/interfaces-log-observe/task.yaml:
```diff
@@ -44,3 +89,30 @@ execute: |
+        exit 0
+    fi
+
+    log-observe-consumer.klogctl SYSLOG_ACTION_READ_ALL | MATCH 'ERROR (operation not permitted|permission denied)'
+    log-observe-consumer.klogctl SYSLOG_ACTION_SIZE_BUFFER | MATCH 'ERROR (operation not permitted|permission denied)'
```
and EPERM 'Operation not permitted' since it's on the implicit deny path now.

bboozzoo reviewed (commented):
Thank you, this looks great. Just some small comments about the spread test.

alexclewontin commented on tests/lib/prepare-restore.sh:
```diff
@@ -609,6 +609,9 @@ prepare_project() {
 
+    # Build binary used for log-observe test (extremely difficult to migrate
+    # this to python or some other interpreted language, as we need to make direct syscalls)
+    go build -o ./tests/lib/snaps/log-observe-consumer/bin ./tests/lib/snaps/log-observe-consumer/src/klogctl
```
Will that work when running the test on Core platforms?

bboozzoo replied:
Ah, good point. Fair enough, let's keep it this way. Easier than building and publishing a snap in the store.

alexclewontin replied:
Sounds good, I'll add a comment here documenting the why here

alexclewontin committed:
tests: narrow allowed errors when testing syslog(2) calls

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

bboozzoo reviewed (approved):
Thank you

zyga commented on tests/lib/snaps/log-observe-consumer/src/klogctl/main.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func main() {
+
```
Spurious newline

zyga commented on tests/lib/snaps/log-observe-consumer/src/klogctl/main.go:
```diff
@@ -0,0 +1,77 @@
+
+	cmdNo, ok := cmdResolver[cmd]
+	if !ok {
+		panic("klogctl: unknown SYSLOG_ACTION_* (see `man 2 syslog`)")
```
Perhaps log.Fatalf instead of a panic?

zyga commented on tests/lib/snaps/log-observe-consumer/src/klogctl/main.go:
```diff
@@ -0,0 +1,77 @@
+func main() {
+
+	if len(os.Args) != 2 {
+		panic("usage: klogctl SYSLOG_ACTION_*")
```
Perhaps log.Fatalf instead of a panic?

zyga commented on tests/lib/snaps/log-observe-consumer/src/klogctl/main.go:
```diff
@@ -0,0 +1,77 @@
+	if err == nil {
+		fmt.Println("SUCCESS")
+	} else {
+		fmt.Printf("ERROR %s\n", err)
```
I think you want this to fail the process too. Converting it to log.Fatalf should be fine here.

alexclewontin replied:
I just reworked the if/else statement and added an os.Exit(1) here because using log.Fatalf made the piping/MATCHing in the spread test more complicated, but I can rework that if you think it's really better

zyga reviewed (changes requested):
Overall +1 and very impressive filtering and testing. I have one small request in the go test binary though. Please ping me for re-review.

alexclewontin committed:
tests: slight tweak to klogctl binary

Signed-off-by: Alex Lewontin <alex.lewontin@canonical.com>

zyga reviewed (approved):
Thank you, this looks great.

pedronis commented on interfaces/seccomp/template.go:
```diff
@@ -543,7 +543,23 @@ fdatasync
+# been set to 1 by default, which requires CAP_SYSLOG for access to /dev/kmsg.
+# Thus, we have to grant CAP_SYSLOG to the log-observe interface, so we must
+# explicitly deny the non-"observe" type privileged accesses here. The following
+# only allows SYSLOG_ACTION_READ{_ALL}, SYSLOG_ACTION_SIZE_{UNREAD,BUFFER}
```
I wonder whether we still need the "capability syslog" in kernel_module_observe.go ? cc @alexmurray

alexmurray replied:
Hmm this is a good point @pedronis - yes I think that with this new reworking of log_observe we can remove this `capability syslog` from kernel-module-observe and instead then just have snaps plug log-observe as well (if they don't already) for that purpose. Or I wonder also if we should have similar seccomp rules in place in kernel-module-observe as well to avoid granting more privileged access than is required?

pedronis reviewed (commented):
my main remaining question is how the filtering will interact with snaps using interfaces that were giving cap syslog already (mostly kernel-module-*), probably low risk that; and snaps that get themselves partially out of confinement, do those manipulate the seccomp filters as well? will they be stopped to do things they were able to do so far?

pedronis commented:
@alexmurray am I confused? the new extra filtering is done in the main seccomp template afaict, not only for log-observe, so choosing interfaces wouldn't change things, no?

alexmurray commented:
@pedronis yes - the base template now contains the explicit deny rules so you are right - I wonder what would happen if we added explicit allow rules to these other interfaces, whether they would override the deny rules from the base template?

pedronis commented:
@alexmurray 

> @pedronis yes - the base template now contains the explicit deny rules so you are right - I wonder what would happen if we added explicit allow rules to these other interfaces, whether they would override the deny rules from the base template?

my memory is that denies wins with how things are organized?

ernestl commented:
Decided with @pedronis to hold back this PR from 2.66, to allow for further thinking/testing.

ernestl commented:
Still require more thinking, moving out of 2.67.

bboozzoo commented on tests/main/interfaces-log-observe/task.yaml:
```diff
@@ -32,10 +45,43 @@ execute: |
+    log-observe-consumer.klogctl SYSLOG_ACTION_CONSOLE_LEVEL | MATCH 'ERROR permission denied'
+    # It is important that the following matches EACCES, not EINVAL, as it
+    # is a stand-in for potential future SYSLOG_ACTIONs, which should be denied by default
+    log-observe-consumer.klogctl SYSLOG_ACTION_BAD | MATCH 'ERROR permission denied'
```
and you should be able to use log.Fatalf() now

```suggestion
    log-observe-consumer.klogctl SYSLOG_ACTION_BAD 2>&1| MATCH 'ERROR permission denied'
```

bboozzoo commented:
> @alexmurray
> 
> > @pedronis yes - the base template now contains the explicit deny rules so you are right - I wonder what would happen if we added explicit allow rules to these other interfaces, whether they would override the deny rules from the base template?
> 
> my memory is that denies wins with how things are organized?

That is my understanding as well, the more restrictive profile will win. We apply allow/deny profiles independently, in which case the deny rules should take priority. This could be different if all rules were in a single profile, and AFAIU it would depend on how they are added/ordered to libseccomp.

pedronis commented:
so the interfaces that gave already syslog capability were:

kernel_module_control.go
kernel_module_observe

I'm quite sure that we don't want 'write' actions for the latter, and it's questionable why one would think to get control on syslog from the former.

We need to be ready introducing a syslog-control interface if somebody needs the 'write' parts that we are blocking here though, we would need to undo the deny bits, in similar ways we deal some conflicting rules with apparmor atm.

So from that POV I think we are ok here. I have a different wondering though that I need to address with @alexmurray

jslarraz reviewed (commented):
Hey folks,

I guess I'm missing something, but is there any reason to specify the deny rules rather than only the allow rules?

It looks cleaner to me, and it would make easier to implement the log_control interface if needed.

alexmurray commented:
Regarding the question from @pedronis on possible conflict between this proposed new interface (with explicit deny seccomp rules) and other interfaces which may want to purposefully grant these same accesses, perhaps this can be alleviated by changing the direction here to instead use the explicit allow rules for only a subset of actions in the base template as suggested by @jslarraz and removing the blanket allow for the `syslog` syscall in the base template - then other more privileged interfaces can simply grant additional actions and there should be no conflict.

@alexclewontin any appetite for (unfortunately) reworking this as suggested by @jslarraz above?

ndyer commented:
This is likely at risk for 2.68 which we are intending to cut next Monday.

ernestl commented:
Moved to 2.69, the task is not ready and it as discussed on Friday 7 Feb, getting it ready in time for 2.68 is not viable.

alexclewontin commented:
@alexmurray Sorry for delay. I can eventually rework, but not on any predictable timeline.
